radix-engine/src/system/system.rs

```diff
@@ -665,24 +582,200 @@ where
+            &substate_key,
+            flags,
+            Some(|| IndexedScryptoValue::from_typed(&Option::<ScryptoValue>::None)),
+        )
+    }
 }
 
-impl<'a, Y, V> ClientSortedStoreApi<RuntimeError> for SystemDownstream<'a, Y, V>
+impl<'a, Y, V> ClientIndexApi<RuntimeError> for SystemDownstream<'a, Y, V>
 where
-    Y: KernelApi<SystemCallback<V>>,
+    Y: KernelApi<SystemConfig<V>>,
+    V: SystemCallbackObject,
+{
+    fn new_index(&mut self) -> Result<NodeId, RuntimeError> {
+        let entity_type = EntityType::InternalIndex;
+        let node_id = self.api.kernel_allocate_node_id(entity_type)?;
+
+        self.api.kernel_create_node(
+            node_id,
+            btreemap!(
+                SysModuleId::Object.into() => btreemap!(),
+                SysModuleId::TypeInfo.into() => ModuleInit::TypeInfo(
+                    TypeInfoSubstate::IterableStore
+                ).to_substates(),
+            ),
+        )?;
+
+        Ok(node_id)
+    }
+
+    fn insert_into_index(
+        &mut self,
+        node_id: &NodeId,
+        key: Vec<u8>,
+        buffer: Vec<u8>,
+    ) -> Result<(), RuntimeError> {
+        let type_info = TypeInfoBlueprint::get_type(&node_id, self.api)?;
+        match type_info {
+            TypeInfoSubstate::IterableStore => {}
+            _ => {
+                return Err(RuntimeError::SystemError(SystemError::NotAnIterableStore));
+            }
+        }
+
+        let value = IndexedScryptoValue::from_vec(buffer).map_err(|e| {
+            RuntimeError::SystemUpstreamError(SystemUpstreamError::InputDecodeError(e))
+        })?;
+
+        if !value.owned_node_ids().is_empty() {
+            return Err(RuntimeError::SystemError(
+                SystemError::CannotStoreOwnedInIterable,
+            ));
+        }
+
+        let module_id = SysModuleId::Object.into();
+        let substate_key = SubstateKey::Map(key);
+
+        self.api
+            .kernel_set_substate(node_id, module_id, substate_key, value)
+    }
+
+    fn remove_from_index(
+        &mut self,
+        node_id: &NodeId,
+        key: Vec<u8>,
+    ) -> Result<Option<Vec<u8>>, RuntimeError> {
+        let type_info = TypeInfoBlueprint::get_type(&node_id, self.api)?;
+        match type_info {
+            TypeInfoSubstate::IterableStore => {}
+            _ => {
+                return Err(RuntimeError::SystemError(SystemError::NotAnIterableStore));
+            }
+        }
+
+        let module_id = SysModuleId::Object.into();
+        let substate_key = SubstateKey::Map(key);
+
+        let rtn = self
+            .api
+            .kernel_remove_substate(node_id, module_id, &substate_key)?
+            .map(|v| v.into());
+
+        Ok(rtn)
+    }
+
+    fn scan_index(&mut self, node_id: &NodeId, count: u32) -> Result<Vec<Vec<u8>>, RuntimeError> {
+        let type_info = TypeInfoBlueprint::get_type(&node_id, self.api)?;
+        match type_info {
+            TypeInfoSubstate::IterableStore => {}
```

Comment:
Should we rename all `IterableStore` to `Index`, or the other way, for consistency?

Reply:
Yup good catch